src/particle_filter.cpp

```diff
@@ -39,65 +101,103 @@ bool compareSamples(const Sample& a, const Sample& b)
+    // Build up cumulative probability table for resampling.
+    // TODO: Replace this with a more efficient procedure
+    // (e.g., http://www.network-theory.co.uk/docs/gslref/GeneralDiscreteDistributions.html)
+    std::vector<double> c;
+    c.resize(old_samples.size() + 1, 0);
+    for (uint i=0; i<old_samples.size(); ++i)
+        c[i+1] = c[i]+ old_samples[i].weight;
 
-        int l = std::min<int>(k + 1 + old_sample.weight * num_samples, num_samples - 1);
+    double w_diff = 1 - w_fast_ / w_slow_;
+    if(w_diff < 0)
+        w_diff = 0;
 
-        for(int i = k; i <= l; ++i)
-            new_samples[i] = old_sample;
+    // Create the kd tree for adaptive sampling;
+    kd_tree_->clear();
 
-        k = l + 1;
+    // Draw samples from set a to create set b.
+    new_samples.clear();
 
-        if (k >= num_samples)
+    while(new_samples.size() < max_samples_)
+    {
+        new_samples.push_back(Sample());
+        Sample& new_sample = new_samples.back();
+
+        double r = drand48();
+
+        if(r < w_diff)
+            new_sample.pose.set(gen_random_pose_function());
+        else
+        {
+            // Naive discrete event sampler
+            uint i=0;
+            for(; i<old_samples.size(); ++i)
+            {
+                if((c[i] <= r) && (r < c[i+1]))
+                    break;
+            }
+
+            // Add sample to list
+            new_sample.pose =  old_samples[i].pose;
+        }
+
+        new_sample.weight = 1;
+
+        // Add sample to histogram
+        kd_tree_->insert(new_sample.pose, new_sample.weight);
+
+        // See if we have enough samples yet
+        if (new_samples.size() >= resampleLimit(kd_tree_->getLeafCount()))
             break;
     }
 
-//    // Build up cumulative probability table for resampling.
-//    std::vector<double> cum_weights(old_samples.size());
-//    cum_weights[0] = old_samples[0].weight;
-//    for(unsigned int i = 1; i < old_samples.size(); ++i)
-//        cum_weights[i] = cum_weights[i - 1] + old_samples[i].weight;
+    switchSamples();
 
-//    new_samples.resize(num_samples);
-//    for(std::vector<Sample>::iterator it = new_samples.begin(); it != new_samples.end(); ++it)
-//    {
-//        Sample& new_sample = *it;
+    normalize();
+}
 
-//        double r = ((double) rand() / (RAND_MAX));
+// ----------------------------------------------------------------------------------------------------
 
-//        unsigned int i_sample = 0;
-//        for(i_sample = 0; i_sample < old_samples.size(); ++i_sample)
-//        {
-//            if (r < cum_weights[i_sample])
-//                break;
-//        }
+unsigned int ParticleFilter::resampleLimit(unsigned int k)
+{
+    if (limit_cache_[k-1] != 0)
+        return limit_cache_[k-1];
 
-//        new_sample = old_samples[i_sample];
-//    }
+    if (k <= 1)
+    {
+        limit_cache_[k-1] = max_samples_;
+        return max_samples_;
+    }
 
-//    new_samples[0] = best_sample;
+    // double a = 1;
```

Comment:
Remove?

Reply:
I prefer to keep it as it is part of the algorithm,  but variable is not needed in coding.